src/hotspot/cpu/aarch64/c1_LIRGenerator_aarch64.cpp

```diff
@@ -277,18 +277,20 @@ void LIRGenerator::cmp_reg_mem(LIR_Condition condition, LIR_Opr reg, LIR_Opr bas
 
 
 bool LIRGenerator::strength_reduce_multiply(LIR_Opr left, jint c, LIR_Opr result, LIR_Opr tmp) {
-
-  if (is_power_of_2(c - 1)) {
-    __ shift_left(left, exact_log2(c - 1), tmp);
+  juint u_value = (juint)c;
+  if (is_power_of_2(u_value - 1)) {
```

Comment:
And do you have some sort of tests for this, to make sure we check with all possible `c` values?

Reply:
> What happens if this underflows? Is this not undefined behaviour? Could we use `java_add`?

Unsigned subtraction is never undefined. "wrap around" behavior is used. Using `java_add` / `java_subtract` sounds like a good idea.

Reply:
>And do you have some sort of tests for this, to make sure we check with all possible c values?

No, as of now I only ran tier1 test cases with c1 compiler. Nothing else.

Reply:
All that `java_add` does is cast to unsigned and then add. That's equivalent to what we're doing here, but explicit casts make the arithmetic clearer, IMO.

Reply:
Ok, I forgot that unsigned has properly defined overflow semantics.

> No, as of now I only ran tier1 test cases with c1 compiler. Nothing else.

Tests would really be great. Is that possible?

Reply:
> Tests would really be great. Is that possible?

I could try, Wouldn't this be already tested by `test/hotspot/jtreg/compiler/c1/MultiplyByMaxInt.java` and `test/hotspot/jtreg/compiler/integerArithmetic/MultiplyByIntegerMinHang.java` tests :)

Reply:
Shouldn't we get an incorrect result with these tests then?

Reply:
> Shouldn't we get an incorrect result with these tests then?

I don't see any failures, again why would there be an incorrect result ? I didn't get that.

Reply:
Is this just undefined behaviour, but no compiler so far actually does something unexpected? If so, it will be impossible to have a failing regression test before the patch. But if there is actually an overflow bug, then there could be a regression test that would be failing before the patch, and we should try to find it.

Reply:
> Is this just undefined behaviour, but no compiler so far actually does something unexpected?

Exactly. The compilers already seem to generate code which matches the unsigned behavior. Only UBSan checks found the issue. They run into the issue with existing tests.

Reply:
Okay, thanks for the clarification!

Reply:
> Tests would really be great. Is that possible?

@eme64 I have added one testcase. Can you have a look.